util/src/main/java/io/kubernetes/client/informer/SharedInformerFactory.java

```diff
@@ -0,0 +1,106 @@
+
+  private Map<Type, SharedIndexInformer> informers;
+
+  private Map<Type, Future> startedInformers;
+
+  private ExecutorService informerExecutor;
+
+  /** Constructor. */
+  public SharedInformerFactory() {
+    informers = new ConcurrentHashMap<>();
+    startedInformers = new HashMap<>();
+    informerExecutor = Executors.newCachedThreadPool();
+  }
+
+  public <ApiType, ApiListType> SharedIndexInformer<ApiType> sharedIndexInformerFor(
+      Function<CallGeneratorParams, Call> callGenerator,
+      Class<ApiType> apiTypeClass,
+      Class<ApiListType> apiListTypeClass) {
+    return sharedIndexInformerFor(callGenerator, apiTypeClass, apiListTypeClass, 0);
+  }
+
+  public <ApiType, ApiListType> SharedIndexInformer<ApiType> sharedIndexInformerFor(
+      Function<CallGeneratorParams, Call> callGenerator,
+      Class<ApiType> apiTypeClass,
+      Class<ApiListType> apiListTypeClass,
+      long resyncPeriodInMillis) {
+    ListerWatcher<ApiType, ApiListType> listerWatcher =
+        listerWatcherFor(callGenerator, apiTypeClass, apiListTypeClass);
+    SharedIndexInformer<ApiType> informer =
+        new DefaultSharedIndexInformer<ApiType, ApiListType>(
+            apiTypeClass, listerWatcher, resyncPeriodInMillis);
+    this.informers.put(TypeToken.get(apiTypeClass).getType(), informer);
+    return informer;
+  }
+
+  public <ApiType, ApiListType> ListerWatcher<ApiType, ApiListType> listerWatcherFor(
+      Function<CallGeneratorParams, Call> callGenerator,
+      Class<ApiType> apiTypeClass,
+      Class<ApiListType> apiListTypeClass) {
+    ApiClient apiClient = Configuration.getDefaultApiClient();
+    return new ListerWatcher<ApiType, ApiListType>() {
+      @Override
+      public ApiListType list(CallGeneratorParams params) throws ApiException {
+        Call call = callGenerator.apply(params);
+        return (ApiListType) apiClient.execute(call, apiListTypeClass).getData();
+      }
+
+      @Override
+      public Watch<ApiType> watch(CallGeneratorParams params) throws ApiException {
+        Call call = callGenerator.apply(params);
+        return Watch.createWatch(
+            apiClient,
+            call,
+            TypeToken.getParameterized(Watch.Response.class, apiTypeClass).getType());
+      }
+    };
+  }
+
+  // Start initializes all requested informers.
+  public synchronized void startAllRegisteredInformers() {
+    if (Collections.isEmptyMap(informers)) {
+      return;
+    }
+    informers.forEach(
+        (informerType, informer) -> {
+          Future future = startedInformers.get(informerType);
+          if (future == null) {
+            startedInformers.put(informerType, informerExecutor.submit(informer::run));
+          }
+        });
+  }
+
+  public synchronized void stopAllRegisteredInformers() {
+    if (Collections.isEmptyMap(informers)) {
+      return;
+    }
+    informers.forEach(
+        (informerType, informer) -> {
+          Future future = startedInformers.get(informerType);
```

Comment:
Same here use `containsKey` Also, I think you want to use `HashMap.remove(...)` not set the type to null. Setting the type to null won't empty the hashmap.

Reply:
nice catch 👍🏻done